cpp/ql/lib/semmle/code/cpp/models/implementations/Gets.qll

```diff
@@ -54,17 +53,60 @@ private class GetsFunction extends DataFlowFunction, TaintFunction, ArrayFunctio
   }
 
   override predicate hasArrayWithVariableSize(int bufParam, int countParam) {
-    not this.hasName("gets") and
     bufParam = 0 and
     countParam = 1
   }
 
+  override predicate hasArrayOutput(int bufParam) { bufParam = 0 }
+
+  override predicate hasSocketInput(FunctionInput input) { input.isParameter(2) }
+}
+
+
+/**
+ * The standard functions `gets`.
+ */
+private class GetsFunction extends DataFlowFunction, TaintFunction, ArrayFunction, AliasFunction,
+  SideEffectFunction, LocalFlowSourceFunction {
+  GetsFunction() {
+    // gets(str)
+    this.hasGlobalOrStdOrBslName("gets")
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
+    output.isReturnValue()
```

Comment:
Hm, shouldn't there also be a case like:
```ql
input.isParameterDeref(0) and
output.isReturnValueDeref(0)
```
to model the fact that we also propagate the pointed-to value to the returned pointed-to value?

(I see that we don't do that for the original model either, so that's probably _also_ a follow-up to this PR).

Reply:
It depends whether we're thinking about the `char *` pointer as a pointer to a buffer or a pointer to a string.  If the latter (and there isn't an error), any existing string is overwritten with new data from `stdin` so there's no data flow.

Reply:
Oh, yes. You're right 👍.